challenge_03/base_solution.py

```diff
@@ -0,0 +1,97 @@
+import typing
+from typing import Tuple, Union
+
+import torch
+import torch.nn as nn
+
+
+class GraphProd2Vec(nn.Module):
+    """
+    p-companion - main stage section 4.1
+    https://assets.amazon.science/d5/16/3f7809974a899a11bacdadefdf24/p-companion-a-principled-framework-for-diversified-complementary-product-recommendation.pdf
+    """
+
+    def __init__(self, in_embedding_len: int, out_embedding_len: int):
+        super().__init__()
+        self.in_embedding_len = in_embedding_len
+        self.out_embedding_len = out_embedding_len
+        self.encoder = self._FFN_encoder()
+
+    def _FFN_encoder(self) -> nn.Module:
+        w_1 = nn.Linear(self.in_embedding_len, self.in_embedding_len)
+        bn_1 = nn.LayerNorm(self.in_embedding_len)
+        tanh_1 = nn.Tanh()
+        w_2 = nn.Linear(self.in_embedding_len, self.in_embedding_len)
+        bn_2 = nn.LayerNorm(self.in_embedding_len)
+        tanh_2 = nn.Tanh()
+        w_3 = nn.Linear(self.in_embedding_len, self.out_embedding_len)
+        bn_3 = nn.LayerNorm(self.out_embedding_len)
+        return nn.Sequential(w_1, bn_1, tanh_1, w_2, bn_2, tanh_2, w_3, bn_3)
+
+    def _GAT(self, emb: torch.tensor, edge: torch.tensor) -> torch.tensor:
+        # should add test on matrix positive semi-definite
+        batch_dim = emb.size(dim=0)
+        edge_dim = edge.size(dim=1)
+        e = torch.exp(torch.matmul(emb, torch.t(emb)))
```

Comment:
neděláme tady zbytečné výpočet všech embeddings?

Reply:
ak to mu správne rozumiem pre každý embd potrebuješ kombináciu ostatných aby si spočítal attention. Alebo som nepochopil otázke

Reply:
tuhle complexitu myslím 
<img width="581" alt="image" src="https://user-images.githubusercontent.com/60660169/224550801-36350b84-1d8f-473c-9336-9c083a9d12dd.png">

Reply:
zjednoduší to i ty následující výpočty

Reply:
ve většině případech grafů se obejdeš bez for-loopů, snaž se je vypustit

Reply:
z optimalizujem to